diff --git a/invoke/runners.py b/invoke/runners.py
index d94a603..6c12600 100644
--- a/invoke/runners.py
+++ b/invoke/runners.py
@@ -506,7 +506,10 @@ class Runner(object):
         if six.PY2:
             # TODO: split up self.encoding, only use the one for 'local
             # encoding' here.
-            string = string.encode(self.encoding)
+            try:
+                string = string.encode(self.encoding)
+            except UnicodeEncodeError as e:
+                string = string.encode('utf-8')
         stream.write(string)
         stream.flush()
 
